node-repository/src/main/java/com/yahoo/vespa/hosted/provision/NodeRepository.java

```diff
@@ -468,24 +469,26 @@ private Node move(Node node, Node.State toState, Agent agent, Optional<String> r
 
     /**
      * Removes a node. A node must be in a legal state before it can be removed.
-     *
-     * @return true if the node was removed, false if it was not found in one of the legal states
      */
-    public boolean remove(String hostname) {
-
-        Node.State[] legalStates = {Node.State.provisioned, Node.State.failed, Node.State.parked};
-        Node.State[] legalDynamicStates = {Node.State.provisioned, Node.State.failed, Node.State.parked, Node.State.dirty};
-
-        Optional<Node> nodeToRemove = getNode(hostname, dynamicAllocationEnabled() ? legalDynamicStates : legalStates);
-        if ( ! nodeToRemove.isPresent()) return false;
+    public void remove(String hostname) {
+        Node nodeToRemove = getNode(hostname).orElseThrow(() ->  new NotFoundException("No node with hostname \"" + hostname + '"'));
+        List<Node.State> legalStates = dynamicAllocationEnabled() ?
+                Arrays.asList(Node.State.provisioned, Node.State.failed, Node.State.parked, Node.State.dirty) :
+                Arrays.asList(Node.State.provisioned, Node.State.failed, Node.State.parked);
+
+        if (! legalStates.contains(nodeToRemove.state())) {
+            throw new IllegalArgumentException("Can only remove node from following states: " +
+                    legalStates.stream().map(Node.State::name).collect(Collectors.joining(", ")));
+        }
 
-        // Only docker nodes are allowed to be deleted in state dirty.
-        if ( nodeToRemove.get().state().equals(Node.State.dirty)) {
-            if (!(nodeToRemove.get().flavor().getType().equals(Flavor.Type.DOCKER_CONTAINER))) return false;
+        if (nodeToRemove.state().equals(Node.State.dirty)) {
+            if (!(nodeToRemove.flavor().getType().equals(Flavor.Type.DOCKER_CONTAINER))) {
+                throw new IllegalArgumentException("Only docker nodes can be deleted from state dirty");
+            }
         }
 
-        try (Mutex lock = lock(nodeToRemove.get())) {
-            return db.removeNode(nodeToRemove.get().state(), hostname);
+        try (Mutex lock = lock(nodeToRemove)) {
+            db.removeNode(nodeToRemove.state(), hostname);
```

Comment:
MInor: Change signature of the method called to be void as well?

Reply:
Changed